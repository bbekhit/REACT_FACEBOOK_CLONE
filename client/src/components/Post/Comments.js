import React, { useEffect, useState } from "react";
import {
  getPosts,
  addComment,
  deleteComment
} from "../../redux/actions/postActions";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { makeStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import Avatar from "@material-ui/core/Avatar";
import Typography from "@material-ui/core/Typography";
import Spinner from "../Spinner/Spinner";
import DeleteOutlinedIcon from "@material-ui/icons/DeleteOutlined";
import EditOutlinedIcon from "@material-ui/icons/EditOutlined";
import { TextField } from "@material-ui/core";

const useStyles = makeStyles(theme => ({
  root: {
    width: "100%",
    backgroundColor: theme.palette.common.whiteColor
  },
  inline: {
    display: "inline",
    marginRight: "1rem"
  },
  wrapper: {
    width: "70vw",
    [theme.breakpoints.down("sm")]: {
      width: "90vw"
    },
    padding: "1rem",
    overflow: "hidden",
    margin: "1rem auto",
    ...theme.customBorder
  },
  btnContainer: {
    [theme.breakpoints.down("sm")]: {
      opacity: 1
    },
    opacity: 0
  },
  followBtn: {
    fontSize: ".5rem",
    textTransform: "uppercase",
    color: theme.palette.common.mainBlue,
    borderStyle: "solid",
    borderWidth: "1px",
    borderColor: theme.palette.common.mainBlue
  }
}));

const Comments = ({
  post,
  getPosts,
  addComment,
  deleteComment,
  auth: { user }
}) => {
  const [comment, setComment] = useState("");
  const onChange = e => {
    setComment(e.target.value);
  };
  const classes = useStyles();
  useEffect(() => {
    getPosts();
  }, [getPosts]);
  return (
    <Grid container direction="column" spacing={2}>
      <Grid item>
        {post ? (
          <Grid container className={classes.wrapper}>
            <Grid item>
              <ListItem alignItems="flex-start" sm={8}>
                <ListItemAvatar>
                  <Avatar alt={post.postedBy} src="" />
                </ListItemAvatar>
                <ListItemText
                  primary={post.title}
                  secondary={
                    <React.Fragment>
                      <Typography
                        component="span"
                        variant="body2"
                        className={classes.inline}
                        color="textPrimary"
                      >
                        {post.postedBy}
                      </Typography>
                      {post.body}
                    </React.Fragment>
                  }
                />
              </ListItem>
            </Grid>
          </Grid>
        ) : (
          <Spinner />
        )}
      </Grid>
      <Grid
        item
        className={classes.wrapper}
        container
        direction="column"
        spacing={2}
      >
        <TextField
          label="Comment"
          variant="outlined"
          id="comment"
          fullWidth
          onChange={onChange}
          value={comment}
          rows={5}
          style={{ marginBottom: ".5rem" }}
        ></TextField>
        <Button
          onClick={() => {
            addComment(comment, post._id);
            setComment("");
          }}
        >
          Add Comment
        </Button>
      </Grid>
      <Grid className={classes.wrapper} item>
        <List className={classes.root}>
          {post && post.comments.length !== 0
            ? post.comments.map((item, i) => (
                <Grid
                  container
                  direction="column"
                  className={classes.list}
                  key={`${item.name}-${i}`}
                >
                  <Grid item>
                    <ListItem alignItems="center" sm={8}>
                      <ListItemAvatar>
                        <Avatar alt={item.postedBy} src="" />
                      </ListItemAvatar>
                      <ListItemText
                        primary={item.title}
                        secondary={
                          <React.Fragment>
                            <Typography
                              component="span"
                              variant="body2"
                              className={classes.inline}
                              color="textPrimary"
                            >
                              {item.postedBy}
                            </Typography>
                            {item.body}
                          </React.Fragment>
                        }
                      />
                    </ListItem>
                  </Grid>
                  <Grid
                    container
                    justify="flex-end"
                    style={{ paddingBottom: ".25rem" }}
                  >
                    <Grid item>
                      {user._id === item.user ? (
                        <Grid item container className={classes.btnContainer}>
                          <Grid
                            container
                            justify="flex-end"
                            style={{ cursor: "pointer" }}
                          >
                            <Grid
                              item
                              style={{ color: "blue" }}
                              component={Link}
                              to={`/edit/${item._id}`}
                            >
                              <EditOutlinedIcon />
                            </Grid>
                            <Grid
                              item
                              style={{ color: "red" }}
                              onClick={() => deleteComment(post._id, item._id)}
                            >
                              <DeleteOutlinedIcon />
                            </Grid>
                          </Grid>
                        </Grid>
                      ) : null}
                    </Grid>
                  </Grid>
                </Grid>
              ))
            : "NO COMMENTS YET"}
        </List>
      </Grid>
    </Grid>
  );
};

const mapStateToProps = (state, props) => ({
  post: state.post.find(item => item._id === props.match.params.id),
  auth: state.auth
});
export default connect(mapStateToProps, {
  getPosts,
  addComment,
  deleteComment
})(Comments);
