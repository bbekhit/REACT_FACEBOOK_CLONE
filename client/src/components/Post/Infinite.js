import React, { useState, useEffect } from "react";

export default function Infinite(props) {
  const [loadMore, setLoadMore] = useState(true);
  const [state, setState] = useState([]);

  useEffect(() => {
    getData(loadMore);
    setLoadMore(false);
  }, [loadMore]);

  useEffect(() => {
    const list = document.getElementById("list");
    if (props.scrollable) {
      // list has fixed height
      list.addEventListener("scroll", e => {
        const el = e.target;
        if (el.scrollTop + el.clientHeight === el.scrollHeight) {
          setLoadMore(true);
        }
      });
    } else {
      // list has auto height
      window.addEventListener("scroll", () => {
        if (
          window.scrollY + window.innerHeight ===
          list.clientHeight + list.offsetTop
        ) {
          setLoadMore(true);
        }
      });
    }
  }, []);

  useEffect(() => {
    const list = document.getElementById("list");

    if (list.clientHeight <= window.innerHeight && list.clientHeight) {
      setLoadMore(true);
    }
  }, [state]);

  const getData = load => {
    if (load) {
      fetch("https://dog.ceo/api/breeds/image/random/1")
        .then(res => {
          return !res.ok ? res.json().then(e => Promise.reject(e)) : res.json();
        })
        .then(res => {
          setState([...state, ...res.message]);
        });
    }
  };

  return (
    <ul id="list">
      {state.map((img, i) => (
        <li style={{ backgroundImage: `url(${img})` }} key={i} />
      ))}
    </ul>
  );
}
