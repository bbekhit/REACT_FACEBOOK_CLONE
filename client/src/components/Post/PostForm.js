import React, { useState } from "react";
import { connect } from "react-redux";
import { makeStyles, useTheme, withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";

const useStyles = makeStyles(theme => ({
  btn: {
    ...theme.typography.submitBtn,
    width: "100%",
    "&:hover": {
      background: theme.palette.common.whiteColor,
      color: theme.palette.common.mainBlue,
      border: "1px solid #4f34ff",
    },
  },
}));

const CssTextField = withStyles(theme => ({
  root: {
    "& label.Mui-focused": {
      color: theme.palette.common.mainBlue,
    },
    "& .MuiInput-underline:after": {
      borderBottomColor: theme.palette.common.mainBlue,
    },
    "& .MuiOutlinedInput-root": {
      "& fieldset": {
        borderColor: theme.palette.common.mainBlue,
      },
      "&:hover fieldset": {
        borderColor: theme.palette.common.mainBlue,
      },
      "&.Mui-focused fieldset": {
        borderColor: theme.palette.common.mainBlue,
      },
    },
  },
}))(TextField);

export const PostForm = ({ onSubmit, auth, post }) => {
  const [values, setValues] = useState({
    title: post ? post.title : "",
    body: post ? post.body : "",
  });
  const [errors, setErrors] = useState({
    titleError: "",
    bodyError: "",
  });
  const { title, body } = values;
  const { titleError, bodyError } = errors;
  const classes = useStyles();
  const theme = useTheme();

  const onChange = event => {
    let valid;
    switch (event.target.id) {
      case "title":
        setValues({ ...values, title: event.target.value });
        valid = /^$/.test(event.target.value);
        if (valid) {
          setErrors({ ...errors, titleError: "Title required" });
        } else {
          setErrors({ ...errors, titleError: "" });
        }
        break;
      case "body":
        setValues({ ...values, body: event.target.value });
        valid = /^$/.test(event.target.value);
        if (valid) {
          setErrors({
            ...errors,
            bodyError: "Post is required",
          });
        } else {
          setErrors({ ...errors, bodyError: "" });
        }
        break;
      default:
        break;
    }
  };

  const onSubmitForm = e => {
    e.preventDefault();
    const postData = {
      ...values,
      postedBy: auth.user.name,
    };
    onSubmit(postData);
  };

  return (
    <div>
      <Grid container>
        <Grid
          item
          container
          direction="column"
          spacing={2}
          style={{ width: "75vw", margin: "0 auto" }}
        >
          <Grid item>
            <CssTextField
              error={!!titleError}
              helperText={titleError}
              className={classes.margin}
              label="Title"
              variant="outlined"
              id="title"
              fullWidth
              InputProps={{
                className: classes.inputColor,
              }}
              InputLabelProps={{
                style: {
                  color: theme.palette.common.mainBlue,
                },
              }}
              onChange={onChange}
              value={title}
              // inputProps={{ style: { fontFamily: 'nunito', color: 'white'}}}
              inputProps={{
                id: "titleInput",
              }}
            />
          </Grid>
          <Grid item>
            <CssTextField
              error={!!bodyError}
              helperText={bodyError}
              className={classes.margin}
              label="Enter your post here..."
              variant="outlined"
              id="body"
              fullWidth
              multiline
              rows={10}
              InputProps={{
                className: classes.inputColor,
              }}
              InputLabelProps={{
                style: {
                  color: theme.palette.common.mainBlue,
                },
              }}
              onChange={onChange}
              value={body}
              // inputProps={{ style: { fontFamily: 'nunito', color: 'white'}}}
            />
          </Grid>
          <Grid item>
            <Button
              className={classes.btn}
              onClick={onSubmitForm}
              disabled={!!titleError || !title || !!bodyError || !body}
              id="btn"
            >
              Submit
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </div>
  );
};

const mapStateToProps = state => ({
  auth: state.auth,
});
export default connect(mapStateToProps)(PostForm);
