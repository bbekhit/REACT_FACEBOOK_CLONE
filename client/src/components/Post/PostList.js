import React, { useEffect, useState } from "react";
import { getPosts, addLike, removeLike } from "../../redux/actions/postActions";
import { openModal } from "../../redux/actions/modalActions";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { makeStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import Grid from "@material-ui/core/Grid";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import Avatar from "@material-ui/core/Avatar";
import Typography from "@material-ui/core/Typography";
import Spinner from "../Spinner/Spinner";
import DeleteOutlinedIcon from "@material-ui/icons/DeleteOutlined";
import EditOutlinedIcon from "@material-ui/icons/EditOutlined";
import ThumbUpAltIcon from "@material-ui/icons/ThumbUpAlt";
import Button from "@material-ui/core/Button";

const useStyles = makeStyles(theme => ({
  root: {
    width: "100%",
    backgroundColor: theme.palette.common.whiteColor
  },
  inline: {
    display: "inline",
    marginRight: "1rem"
  },
  list: {
    ...theme.customBorder,
    borderRadius: "1rem",
    marginBottom: "1rem",
    textDecoration: "none",
    padding: ".5rem",
    color: "black",
    overflow: "hidden",
    "&:hover": {
      cursor: "pointer",
      "& $btnContainer": {
        opacity: 1
      }
    }
  },
  btnContainer: {
    [theme.breakpoints.down("sm")]: {
      opacity: 1
    },
    opacity: 0
  },
  followBtn: {
    fontSize: ".5rem",
    textTransform: "uppercase",
    color: theme.palette.common.mainBlue,
    borderStyle: "solid",
    borderWidth: "1px",
    borderColor: theme.palette.common.mainBlue
  }
}));

const PostList = ({
  getPosts,
  posts,
  addLike,
  removeLike,
  openModal,
  auth: { user }
}) => {
  const [limit, setLimit] = useState(5);
  const [skip, setSkip] = useState(0);
  const [size, setSize] = useState(Infinity);
  const classes = useStyles();

  useEffect(() => {
    getPosts(skip, limit);
  }, [getPosts]);

  const loadMore = async () => {
    let newSkip = skip + limit;
    let size = await getPosts(newSkip, limit, posts);
    setSize(size);
    setSkip(newSkip);
  };

  return (
    <Grid container style={{ padding: "2rem" }} className={classes.wrapper}>
      <List className={classes.root}>
        {posts ? (
          posts.map((item, i) => (
            <Grid
              container
              direction="column"
              className={classes.list}
              key={`${item.name}-${i}`}
            >
              <Grid item>
                <ListItem alignItems="flex-start" sm={8}>
                  <ListItemAvatar>
                    <Avatar alt={item.postedBy} src="" />
                  </ListItemAvatar>
                  <ListItemText
                    primary={item.title}
                    secondary={
                      <React.Fragment>
                        <Typography
                          component="span"
                          variant="body2"
                          className={classes.inline}
                          color="textPrimary"
                        >
                          {item.postedBy}
                        </Typography>
                        {item.body}
                      </React.Fragment>
                    }
                  />
                </ListItem>
              </Grid>
              <Grid
                container
                justify="space-around"
                style={{ paddingBottom: ".25rem" }}
              >
                <Grid item>
                  <Button
                    component={Link}
                    to={`/post/${item._id}`}
                    className={classes.followBtn}
                  >
                    Comments
                  </Button>
                </Grid>
                <Grid item>
                  {user._id === item.user ? (
                    <Grid item container className={classes.btnContainer}>
                      <Grid
                        container
                        justify="flex-end"
                        style={{ cursor: "pointer" }}
                      >
                        <Grid
                          item
                          style={{ color: "blue" }}
                          component={Link}
                          to={`/edit/${item._id}`}
                        >
                          <EditOutlinedIcon />
                        </Grid>
                        <Grid
                          item
                          style={{ color: "red" }}
                          onClick={() =>
                            openModal("ConfirmModal", {
                              isOpen: true,
                              postId: item._id
                            })
                          }
                        >
                          <DeleteOutlinedIcon />
                        </Grid>
                      </Grid>
                    </Grid>
                  ) : null}
                </Grid>
                <Grid item style={{ alignSelf: "center" }}>
                  <ThumbUpAltIcon
                    onClick={() => {
                      item.likes.includes(user._id)
                        ? removeLike(item._id)
                        : addLike(item._id);
                    }}
                    color={
                      item.likes.includes(user._id) ? "primary" : "inherit"
                    }
                  />
                  <span
                    style={{
                      marginLeft: "3px",
                      position: "relative",
                      bottom: "5px"
                    }}
                  >
                    {item.likes.length !== 0 ? item.likes.length : null}
                  </span>
                </Grid>
              </Grid>
            </Grid>
          ))
        ) : (
          <Spinner />
        )}
      </List>
      <Grid container justify="flex-end">
        {size > 0 && size >= limit ? (
          <Button className={classes.followBtn} onClick={loadMore}>
            Load More
          </Button>
        ) : null}
      </Grid>
    </Grid>
  );
};

const mapStateToProps = state => ({
  posts: state.post,
  auth: state.auth
});
export default connect(mapStateToProps, {
  getPosts,
  openModal,
  addLike,
  removeLike
})(PostList);
