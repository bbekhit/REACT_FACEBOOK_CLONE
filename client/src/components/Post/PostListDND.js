import React, { useEffect, useState } from "react";
import { getPosts } from "../../redux/actions/postActions";
import { connect } from "react-redux";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
import uuid from "uuid/v4";
import Spinner from "../Spinner/Spinner";

const PostListDND = ({ posts, getPosts }) => {
  const [limit, setLimit] = useState(5);
  const [skip, setSkip] = useState(0);

  const columnsFromBackend = {
    [uuid()]: {
      name: "Requested",
      items: posts,
    },
    [uuid()]: {
      name: "To do",
      items: [],
    },
    [uuid()]: {
      name: "In Progress",
      items: [],
    },
    [uuid()]: {
      name: "Done",
      items: [],
    },
  };
  const [columns, setColumns] = useState(columnsFromBackend);

  useEffect(() => {
    getPosts(skip, limit);
  }, [getPosts]);

  useEffect(() => {
    setColumns(columnsFromBackend);
  }, [posts]);

  const onDragEnd = (result, columns, setColumns) => {
    if (!result.destination) return;
    const { source, destination } = result;

    if (source.droppableId !== destination.droppableId) {
      const sourceColumn = columns[source.droppableId];
      const destColumn = columns[destination.droppableId];
      const sourceItems = [...sourceColumn.items];
      const destItems = [...destColumn.items];
      const [removed] = sourceItems.splice(source.index, 1);
      destItems.splice(destination.index, 0, removed);
      setColumns({
        ...columns,
        [source.droppableId]: {
          ...sourceColumn,
          items: sourceItems,
        },
        [destination.droppableId]: {
          ...destColumn,
          items: destItems,
        },
      });
    } else {
      // one column
      const column = columns[source.droppableId]; // get the column
      const copiedItems = [...column.items]; // make a copy of the items inside the column
      const [removed] = copiedItems.splice(source.index, 1); // return array of removed item
      console.log("removed", removed);
      console.log("source", source);
      console.log("destination", destination);

      copiedItems.splice(destination.index, 0, removed);
      setColumns({
        ...columns,
        [source.droppableId]: {
          ...column,
          items: copiedItems,
        },
      });
    }
  };

  if (posts.length === 0) {
    return <Spinner />;
  }

  return (
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        width: "100vw",
        margin: "0 auto",
        border: "1px dashed black",
      }}
    >
      <DragDropContext
        onDragEnd={(result) => onDragEnd(result, columns, setColumns)}
      >
        {Object.entries(columns).map(([columnId, column], index) => {
          return (
            <div
              style={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                border: "1px dashed red",
              }}
              key={columnId}
            >
              <h2>{column.name}</h2>
              <div style={{ margin: 8, border: "1px dashed green" }}>
                <Droppable droppableId={columnId}>
                  {(provided, snapshot) => {
                    return (
                      <div
                        {...provided.droppableProps}
                        ref={provided.innerRef}
                        style={{
                          background: snapshot.isDraggingOver
                            ? "lightblue"
                            : "lightgrey",
                          padding: 4,
                          width: 250,
                          minHeight: 500,
                          border: "1px dashed orange",
                        }}
                      >
                        {column.items.map((item, index) => {
                          return (
                            <Draggable
                              key={item._id}
                              draggableId={item._id}
                              index={index}
                            >
                              {(provided, snapshot) => {
                                return (
                                  <div
                                    ref={provided.innerRef}
                                    {...provided.draggableProps}
                                    {...provided.dragHandleProps}
                                    style={{
                                      userSelect: "none",
                                      padding: 16,
                                      margin: "0 0 8px 0",
                                      minHeight: "50px",
                                      backgroundColor: snapshot.isDragging
                                        ? "#263B4A"
                                        : "#456C86",
                                      color: "white",
                                      border: "1px dashed purple",
                                      ...provided.draggableProps.style,
                                    }}
                                  >
                                    {/* {item.content} */}
                                    {item.title}
                                  </div>
                                );
                              }}
                            </Draggable>
                          );
                        })}
                        {provided.placeholder}
                      </div>
                    );
                  }}
                </Droppable>
              </div>
            </div>
          );
        })}
      </DragDropContext>
    </div>
  );
};

const mapStateToProps = (state) => ({
  posts: state.post,
  auth: state.auth,
});
export default connect(mapStateToProps, {
  getPosts,
})(PostListDND);
