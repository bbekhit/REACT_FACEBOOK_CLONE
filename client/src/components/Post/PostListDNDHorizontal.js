import React, { useEffect, useState } from "react";
import { getPosts } from "../../redux/actions/postActions";
import { connect } from "react-redux";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
import uuid from "uuid/v4";
import Spinner from "../Spinner/Spinner";

const PostListDND = ({ posts, getPosts }) => {
  const [limit, setLimit] = useState(4);
  const [skip, setSkip] = useState(0);
  const [homeIndex, setHomeIndex] = useState(null);

  const columnsFromBackend = {
    "column-1": {
      name: "Requested",
      items: posts,
    },
  };
  // one-way
  // const columnOrder = ["column-1", "column-2", "column-3", "column-4"];

  const [columns, setColumns] = useState(columnsFromBackend);

  useEffect(() => {
    getPosts(skip, limit);
  }, [getPosts]);

  useEffect(() => {
    setColumns(columnsFromBackend);
  }, [posts]);

  // one-way
  // const onDragStart = (start) => {
  //   const index = columnOrder.indexOf(start.source.droppableId);
  //   setHomeIndex(index);
  // };

  const onDragEnd = (result, columns, setColumns) => {
    // one-way
    // setHomeIndex(null);
    if (!result.destination) return;
    const { source, destination } = result;

    if (source.droppableId !== destination.droppableId) {
      const sourceColumn = columns[source.droppableId];
      const destColumn = columns[destination.droppableId];
      const sourceItems = [...sourceColumn.items];
      const destItems = [...destColumn.items];
      const [removed] = sourceItems.splice(source.index, 1);
      destItems.splice(destination.index, 0, removed);
      setColumns({
        ...columns,
        [source.droppableId]: {
          ...sourceColumn,
          items: sourceItems,
        },
        [destination.droppableId]: {
          ...destColumn,
          items: destItems,
        },
      });
    } else {
      // one column
      const column = columns[source.droppableId]; // get the column
      const copiedItems = [...column.items]; // make a copy of the items inside the column
      const [removed] = copiedItems.splice(source.index, 1); // return array of removed item
      console.log("removed", removed);
      console.log("source", source);
      console.log("destination", destination);

      copiedItems.splice(destination.index, 0, removed);
      setColumns({
        ...columns,
        [source.droppableId]: {
          ...column,
          items: copiedItems,
        },
      });
    }
  };

  if (posts.length === 0) {
    return <Spinner />;
  }

  return (
    <div
      style={{
        display: "flex",
        justifyContent: "center",
        width: "100vw",
        margin: "0 auto",
        border: "1px dashed black",
      }}
    >
      <DragDropContext
        onDragEnd={(result) => onDragEnd(result, columns, setColumns)}
        // one-way
        // onDragStart={onDragStart}
      >
        {Object.entries(columns).map(([columnId, column], index) => {
          return (
            <div
              style={{
                display: "flex",
                flexDirection: "column",
                alignItems: "center",
                border: "1px dashed red",
              }}
              key={columnId}
            >
              <h2>{column.name}</h2>
              <div style={{ margin: 8, border: "1px dashed green" }}>
                <Droppable
                  droppableId={columnId}
                  isDropDisabled={index < homeIndex}
                  direction="horizontal"
                >
                  {(provided, snapshot) => {
                    return (
                      <div
                        {...provided.droppableProps}
                        ref={provided.innerRef}
                        style={{
                          background: snapshot.isDraggingOver
                            ? "lightblue"
                            : "lightgrey",
                          padding: 4,
                          border: "1px dashed orange",
                          display: "flex",
                          justifyContent: "space-between",
                        }}
                      >
                        {column.items.map((item, index) => {
                          return (
                            <Draggable
                              key={item._id}
                              draggableId={item._id}
                              index={index}
                            >
                              {(provided, snapshot) => {
                                return (
                                  <div
                                    ref={provided.innerRef}
                                    {...provided.draggableProps}
                                    {...provided.dragHandleProps}
                                    style={{
                                      userSelect: "none",
                                      padding: 16,
                                      margin: "0 8px 0 0",
                                      backgroundColor: snapshot.isDragging
                                        ? "#263B4A"
                                        : "#456C86",
                                      color: "white",
                                      border: "1px dashed purple",
                                      borderRadius: "50%",
                                      width: "40px",
                                      height: "40px",
                                      display: "flex",
                                      justifyContent: "center",
                                      alignItems: "center",
                                      ...provided.draggableProps.style,
                                    }}
                                  >
                                    {/* {item.content} */}
                                    {item.title.split("")[0]}
                                  </div>
                                );
                              }}
                            </Draggable>
                          );
                        })}
                        {provided.placeholder}
                      </div>
                    );
                  }}
                </Droppable>
              </div>
            </div>
          );
        })}
      </DragDropContext>
    </div>
  );
};

const mapStateToProps = (state) => ({
  posts: state.post,
  auth: state.auth,
});
export default connect(mapStateToProps, {
  getPosts,
})(PostListDND);
// https://www.youtube.com/watch?v=uHO3mQgs-e8
