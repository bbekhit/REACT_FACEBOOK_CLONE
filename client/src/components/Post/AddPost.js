import React from "react";
import PostForm from "./PostForm";
import { createPost } from "../../redux/actions/postActions";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";

const AddPost = ({ createPost, history, auth: { isAuthenticated } }) => {
  const onSubmitAdd = async post => {
    let result = await createPost(post);
    if (result === "success") {
      history.push("/posts");
    }
  };

  if (!isAuthenticated) {
    return <Redirect to="/login" />;
  }
  return (
    <div style={{ margin: "5rem auto" }}>
      <PostForm onSubmit={post => onSubmitAdd(post)} />
    </div>
  );
};

const mapStateToProps = state => ({
  auth: state.auth
});
export default connect(mapStateToProps, { createPost })(AddPost);
