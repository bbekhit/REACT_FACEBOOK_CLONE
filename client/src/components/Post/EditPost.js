import React, { useEffect } from "react";
import PostForm from "./PostForm";
import { Redirect } from "react-router-dom";
import { connect } from "react-redux";
import { editPost, getPosts } from "../../redux/actions/postActions";

const EditPost = ({
  editPost,
  getPosts,
  post,
  history,
  auth: { isAuthenticated }
}) => {
  useEffect(() => {
    getPosts();
  }, [getPosts]);

  const onSubmitEdit = async data => {
    let result = await editPost(data, post._id);
    if (result === "success") {
      history.push("/posts");
    }
  };

  if (!isAuthenticated) {
    return <Redirect to="/login" />;
  }
  return (
    <div>
      <PostForm onSubmit={data => onSubmitEdit(data)} post={post} />
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  post: state.post.find(item => item._id === props.match.params.id),
  auth: state.auth
});
export default connect(mapStateToProps, { editPost, getPosts })(EditPost);
