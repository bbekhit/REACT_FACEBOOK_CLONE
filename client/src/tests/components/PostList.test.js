import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { PostList } from "../../components/Post/PostList";
import posts from "../mockData/mockData";

describe("Posts List", () => {
  let auth;

  beforeEach(() => {
    auth = {
      user: {
        isAuthenticated: true,
        _id: 1,
        name: "Bonny",
        email: "bonny@bonny.com",
      },
    };
  });

  it("should render a list of posts", () => {
    let wrapper = shallow(<PostList posts={posts} auth={auth} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should render a spinner if no posts found of posts", () => {
    let wrapper = shallow(<PostList posts={[]} auth={auth} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
