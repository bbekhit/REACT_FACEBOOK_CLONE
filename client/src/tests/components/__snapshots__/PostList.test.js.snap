// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts List should render a list of posts 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  style={
    Object {
      "padding": "2rem",
    }
  }
>
  <WithStyles(ForwardRef(List))
    className="makeStyles-root-1"
  >
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-list-3"
      container={true}
      direction="column"
      key="undefined-0"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(ListItem))
          alignItems="flex-start"
          sm={8}
        >
          <WithStyles(ForwardRef(ListItemAvatar))>
            <WithStyles(ForwardRef(Avatar))
              alt="1"
              src=""
            />
          </WithStyles(ForwardRef(ListItemAvatar))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Title-1"
            secondary={
              <React.Fragment>
                <ForwardRef(WithStyles)
                  className="makeStyles-inline-2"
                  color="textPrimary"
                  component="span"
                  variant="body2"
                >
                  1
                </ForwardRef(WithStyles)>
                Body-1
              </React.Fragment>
            }
          />
        </WithStyles(ForwardRef(ListItem))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="space-around"
        style={
          Object {
            "paddingBottom": ".25rem",
          }
        }
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            className="makeStyles-followBtn-5"
            component={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "Link",
                "propTypes": Object {
                  "innerRef": [Function],
                  "onClick": [Function],
                  "replace": [Function],
                  "target": [Function],
                  "to": [Function],
                },
                "render": [Function],
              }
            }
            to="/post/1"
          >
            Comments
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        />
        <WithStyles(ForwardRef(Grid))
          item={true}
          style={
            Object {
              "alignSelf": "center",
            }
          }
        >
          <ThumbUpAltIcon
            color="inherit"
            onClick={[Function]}
          />
          <span
            style={
              Object {
                "bottom": "5px",
                "marginLeft": "3px",
                "position": "relative",
              }
            }
          />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-list-3"
      container={true}
      direction="column"
      key="undefined-1"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(ListItem))
          alignItems="flex-start"
          sm={8}
        >
          <WithStyles(ForwardRef(ListItemAvatar))>
            <WithStyles(ForwardRef(Avatar))
              alt="2"
              src=""
            />
          </WithStyles(ForwardRef(ListItemAvatar))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Title-2"
            secondary={
              <React.Fragment>
                <ForwardRef(WithStyles)
                  className="makeStyles-inline-2"
                  color="textPrimary"
                  component="span"
                  variant="body2"
                >
                  2
                </ForwardRef(WithStyles)>
                Body-2
              </React.Fragment>
            }
          />
        </WithStyles(ForwardRef(ListItem))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="space-around"
        style={
          Object {
            "paddingBottom": ".25rem",
          }
        }
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Button))
            className="makeStyles-followBtn-5"
            component={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "Link",
                "propTypes": Object {
                  "innerRef": [Function],
                  "onClick": [Function],
                  "replace": [Function],
                  "target": [Function],
                  "to": [Function],
                },
                "render": [Function],
              }
            }
            to="/post/2"
          >
            Comments
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        />
        <WithStyles(ForwardRef(Grid))
          item={true}
          style={
            Object {
              "alignSelf": "center",
            }
          }
        >
          <ThumbUpAltIcon
            color="inherit"
            onClick={[Function]}
          />
          <span
            style={
              Object {
                "bottom": "5px",
                "marginLeft": "3px",
                "position": "relative",
              }
            }
          />
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="flex-end"
  >
    <WithStyles(ForwardRef(Button))
      className="makeStyles-followBtn-5"
      onClick={[Function]}
    >
      Load More
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Posts List should render a spinner if no posts found of posts 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  style={
    Object {
      "padding": "2rem",
    }
  }
>
  <WithStyles(ForwardRef(List))
    className="makeStyles-root-1"
  >
    <Spinner />
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="flex-end"
  >
    <WithStyles(ForwardRef(Button))
      className="makeStyles-followBtn-5"
      onClick={[Function]}
    >
      Load More
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
